File: Chess.cs

public interface IMoveable
{
    int Move(int CurrentPostion);
}

public class MoveByOne : IMoveable
{
    public int Move(int CurrentPostion)
    {
        return CurrentPostion += 1;
    }
}

public class MoveByTwo : IMoveable
{
    public int Move(int CurrentPostion)
    {
        return CurrentPostion += 2;
    }
}

public class InheritanceSoldier : MoveByOne
{
    int Position { get; set; }

    public void Move()
    {
        int nextPosition = Move(Position);
        // do anything with nextPosition
    }
}

public class CompositionSoldier
{
    int Position { get; set; }
    IMoveable _moveable;

    public void SetMoveBahavaior(IMoveable moveable) { _moveable = moveable; }

    public void Move()
    {
        int nextPosition = _moveable.Move(Position);
        // do anything with nextPosition
    }
}

public class Soldier : CompositionSoldier
{
    bool someCondition = true;
    MoveByOne moveByOne = new MoveByOne();
    MoveByTwo moveByTwo = new MoveByTwo();

    public void Main()
    {
        SetMoveBahavaior(moveByOne);
        Move();

        if (someCondition) SetMoveBahavaior(moveByTwo);
        Move();
    }
}

File: Chess.php
<?php

interface MoveBehavior {
    public function Move($currentPosition);
}

class MoveByOne implements MoveBehavior {
    public function Move($currentPosition) {return $currentPosition += 1;}
}

class MoveByTwo implements MoveBehavior {
    public function Move($currentPosition) {return $currentPosition += 2;}
}

class InheritanceSoldier extends MoveByOne
{
    public $Position;
    
    function __construct($Position) {
        $this->Position = $Position;
    }
    
    public function main()
    {
        $nextPosition = Move($this->Position);
        // do anything with nextPosition
    }
}

class CompositionSoldier
{
    public $Position;
    protected $_Moveable;

    public function SetMoveBahavaior($Moveable) { $this->_Moveable = $Moveable; }

    public function Move()
    {
        $nextPosition = $this->_Moveable.Move($this->Position);
        // do anything with nextPosition
    }
}

class Soldier extends CompositionSoldier
{
    public $someCondition;
    public $moveByOne;
    public $moveByTwo;
    
    function __construct() {
        $this->someCondition = true;
        
        $moveByOne = new MoveByOne();
        $moveByTwo = new MoveByTwo();
    }
    

    public function Main()
    {
        SetMoveBahavaior($moveByOne);
        Move();

        if ($this->someCondition) SetMoveBahavaior($moveByTwo);
        Move();
    }
}

?>
